// @clerk/backend-core/dist/cjs/__tests__/TestBackendAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestBackendAPIClient = void 0;
const tslib_1 = require("tslib");
const node_fetch_1 = (0, tslib_1.__importDefault)(require("node-fetch"));
const ClerkBackendAPI_1 = require("../api/ClerkBackendAPI");
const defaultApiKey = process.env.CLERK_API_KEY || '';
const defaultApiVersion = process.env.CLERK_API_VERSION || 'v1';
const defaultServerApiUrl = process.env.CLERK_API_URL || 'https://api.clerk.dev';
exports.TestBackendAPIClient = new ClerkBackendAPI_1.ClerkBackendAPI({
    apiKey: defaultApiKey,
    apiVersion: defaultApiVersion,
    serverApiUrl: defaultServerApiUrl,
    libName: 'test',
    libVersion: '0.0.1',
    packageRepo: 'test',
    fetcher: async (url, { method, authorization, contentType, userAgent, body }) => {
        return (await (0, node_fetch_1.default)(url, {
            method,
            headers: {
                Authorization: authorization,
                'Content-Type': contentType,
                'User-Agent': userAgent,
            },
            ...(body && { body: JSON.stringify(body) }),
        })).json();
    },
});
