// @clerk/backend-core/dist/cjs/api/ClerkBackendAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClerkBackendAPI = void 0;
const tslib_1 = require("tslib");
const collection_1 = require("./collection");
const RestClient_1 = (0, tslib_1.__importDefault)(require("./utils/RestClient"));
const defaultApiKey = process.env.CLERK_API_KEY || '';
const defaultApiVersion = process.env.CLERK_API_VERSION || 'v1';
const defaultServerApiUrl = process.env.CLERK_API_URL || 'https://api.clerk.dev';
class ClerkBackendAPI {
    constructor({ apiKey = defaultApiKey, serverApiUrl = defaultServerApiUrl, apiVersion = defaultApiVersion, fetcher, libName, libVersion, packageRepo, }) {
        this._restClient = new RestClient_1.default(apiKey, serverApiUrl, apiVersion, fetcher, libName, libVersion, packageRepo);
    }
    fetchInterstitial() {
        return this._restClient.fetchInterstitial();
    }
    set apiKey(value) {
        this._restClient.apiKey = value;
    }
    set serverApiUrl(value) {
        this._restClient.serverApiUrl = value;
    }
    set apiVersion(value) {
        this._restClient.apiVersion = value;
    }
    get allowlistIdentifiers() {
        if (!this._allowlistIdentifierApi) {
            this._allowlistIdentifierApi = new collection_1.AllowlistIdentifierApi(this._restClient);
        }
        return this._allowlistIdentifierApi;
    }
    get clients() {
        if (!this._clientApi) {
            this._clientApi = new collection_1.ClientApi(this._restClient);
        }
        return this._clientApi;
    }
    get emails() {
        if (!this._emailApi) {
            this._emailApi = new collection_1.EmailApi(this._restClient);
        }
        return this._emailApi;
    }
    get invitations() {
        if (!this._invitationApi) {
            this._invitationApi = new collection_1.InvitationApi(this._restClient);
        }
        return this._invitationApi;
    }
    get sessions() {
        if (!this._sessionApi) {
            this._sessionApi = new collection_1.SessionApi(this._restClient);
        }
        return this._sessionApi;
    }
    get smsMessages() {
        if (!this._smsMessageApi) {
            this._smsMessageApi = new collection_1.SMSMessageApi(this._restClient);
        }
        return this._smsMessageApi;
    }
    get users() {
        if (!this._userApi) {
            this._userApi = new collection_1.UserApi(this._restClient);
        }
        return this._userApi;
    }
}
exports.ClerkBackendAPI = ClerkBackendAPI;
