// @clerk/backend-core/dist/cjs/api/collection/SessionApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SessionApi = void 0;
const AbstractApi_1 = require("./AbstractApi");
class SessionApi extends AbstractApi_1.AbstractApi {
    async getSessionList(queryParams) {
        return this._restClient.makeRequest({
            method: 'GET',
            path: '/sessions',
            queryParams: queryParams,
        });
    }
    async getSession(sessionId) {
        this.requireId(sessionId);
        return this._restClient.makeRequest({
            method: 'GET',
            path: `/sessions/${sessionId}`,
        });
    }
    async revokeSession(sessionId) {
        this.requireId(sessionId);
        return this._restClient.makeRequest({
            method: 'POST',
            path: `/sessions/${sessionId}/revoke`,
        });
    }
    async verifySession(sessionId, token) {
        this.requireId(sessionId);
        return this._restClient.makeRequest({
            method: 'POST',
            path: `/sessions/${sessionId}/verify`,
            bodyParams: { token },
        });
    }
}
exports.SessionApi = SessionApi;
