// @clerk/backend-core/dist/cjs/api/resources/EmailAddress.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmailAddress = void 0;
const tslib_1 = require("tslib");
const Associations_1 = (0, tslib_1.__importDefault)(require("../utils/Associations"));
const Enums_1 = require("./Enums");
const IdentificationLink_1 = require("./IdentificationLink");
const Verification_1 = require("./Verification");
class EmailAddress {
    constructor(data = {}) {
        Object.assign(this, EmailAddress.defaults, data);
    }
    static fromJSON(data) {
        return new EmailAddress({
            id: data.id,
            emailAddress: data.email_address,
            verification: Verification_1.Verification.fromJSON(data.verification),
            linkedTo: (data.linked_to || []).map((link) => {
                return IdentificationLink_1.IdentificationLink.fromJSON(link);
            }),
        });
    }
}
exports.EmailAddress = EmailAddress;
EmailAddress.attributes = ['id', 'emailAddress'];
EmailAddress.associations = {
    verification: Enums_1.Association.HasOne,
    linkedTo: Enums_1.Association.HasMany,
};
EmailAddress.defaults = (0, Associations_1.default)(EmailAddress.associations);
