// @clerk/backend-core/dist/cjs/api/resources/PhoneNumber.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PhoneNumber = void 0;
const tslib_1 = require("tslib");
const Associations_1 = (0, tslib_1.__importDefault)(require("../utils/Associations"));
const Enums_1 = require("./Enums");
const IdentificationLink_1 = require("./IdentificationLink");
const Verification_1 = require("./Verification");
class PhoneNumber {
    constructor(data = {}) {
        Object.assign(this, PhoneNumber.defaults, data);
    }
    static fromJSON(data) {
        return new PhoneNumber({
            id: data.id,
            phoneNumber: data.phone_number,
            reservedForSecondFactor: data.reserved_for_second_factor,
            verification: Verification_1.Verification.fromJSON(data.verification),
            linkedTo: (data.linked_to || []).map((link) => {
                return IdentificationLink_1.IdentificationLink.fromJSON(link);
            }),
        });
    }
}
exports.PhoneNumber = PhoneNumber;
PhoneNumber.attributes = ['id', 'phoneNumber', 'reservedForSecondFactor'];
PhoneNumber.associations = {
    verification: Enums_1.Association.HasOne,
    linkedTo: Enums_1.Association.HasMany,
};
PhoneNumber.defaults = (0, Associations_1.default)(PhoneNumber.associations);
